book/webapps/url_parsing.md

```diff
@@ -142,7 +142,197 @@ So now we can handle URL fragments as well!
 Now that we have seen a few parsers, we should look at how this fits into a `Browser.application` program. Rather than just saving the current URL like last time, can we parse it into useful data and show that instead?
 
 ```elm
-TODO
+module Main exposing (Model, Msg(..), Route(..), init, links, main, routeParser, subscriptions, toRoute, update, view, viewLink)
+
+import Browser
+import Browser.Navigation as Nav
+import Debug
+import Html exposing (..)
+import Html.Attributes exposing (..)
+import Url exposing (Url)
+import Url.Parser exposing ((</>), Parser, int, map, oneOf, parse, s, string, top)
+
+
+
+-- MAIN
+
+
+main : Program () Model Msg
+main =
+    Browser.application
+        { init = init
+        , view = view
+        , update = update
+        , subscriptions = subscriptions
+        , onUrlChange = UrlChanged
+        , onUrlRequest = LinkClicked
+        }
+
+
+
+-- MODEL
+
+
+type alias Model =
+    { key : Nav.Key
+    , route : Route
+    }
+
+
+type Route
+    = Topic String
+    | Blog Int
+    | User String
+    | Comment String Int
+    | NotFound
+    | Home
+
+
+init : () -> Url.Url -> Nav.Key -> ( Model, Cmd Msg )
+init flags url key =
+    ( Model key Home, Cmd.none )
```

Comment:
I've not test yet, I guess that if user want a direct access to ``blog/4`` we have to replace ``Home`` by  ``(toRoute url)`` other wise it would display the home page, am I correct ?

Reply:
oups I was looking at the diff tabs looks like same comment as @laurentpayot !

Reply:
Thanks, fixed!